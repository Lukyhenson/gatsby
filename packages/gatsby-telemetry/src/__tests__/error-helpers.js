const { sanitizeError, cleanPaths } = require(`../error-helpers`)

describe(`Errors Helpers`, () => {
  describe(`sanitizeError`, () => {
    it(`Removes env, output and converts buffers to strings from execa output`, () => {
      const tags = {
        error: [
          {
            error: null,
            cmd: `git commit -m"test"`,
            file: `/bin/sh`,
            args: [`/bin/sh`, `-c`, `git commit -m"test`],
            options: {
              cwd: `here`,
              shell: true,
              envPairs: [`VERSION=1.2.3`],
              stdio: [{}, {}, {}], // pipes
            },
            envPairs: [`VERSION=1.2.3`],

            stderr: Buffer.from(`this is a an error`),
            stdout: Buffer.from(`this is a test`),
          },
        ],
      }

      sanitizeError(tags)
      const error = tags.error[0]
      expect(error.stderr).toStrictEqual(`this is a an error`)
      expect(error.stdout).toStrictEqual(`this is a test`)
      expect(error.envPairs).toBeUndefined()
      expect(error.options).toBeUndefined()
    })

    it(`Sanitizes current path from error stracktraces`, () => {
      const errormessage = `this is a test`
      let e
      try {
        throw new Error(errormessage)
      } catch (error) {
        e = error
      }
      expect(e).toBeDefined()
      expect(e.message).toEqual(errormessage)
      expect(e.stack).toBeDefined()
      const localPathRegex = new RegExp(
        process.cwd().replace(/[-[/{}()*+?.\\^$|]/g, `\\$&`)
      )
      expect(localPathRegex.test(e.stack)).toBeTruthy()
      const tags = { error: [e] }

      sanitizeError(tags)

      expect(e.stack).toMatchSnapshot()
    })
  })
  describe(`cleanPaths`, () => {
    it(`cleans unix paths`, () => {
      const mockPath = `/Users/username/gatsby/packages/gatsby-telemetry`

      const mockCwd = jest
        .spyOn(process, `cwd`)
        .mockImplementation(() => mockPath)

      const errormessage = `this${mockPath}is\n${mockPath}a test: ${mockPath}`

      expect(cleanPaths(errormessage)).toBe(`this$PWDis\n$PWDa test: $PWD`)
      mockCwd.mockRestore()
    })
    it(`cleans Windows paths`, () => {
      const mockPath = `C:\\Users\\username\\gatsby\\packages\\gatsby-telemetry`

      const mockCwd = jest
        .spyOn(process, `cwd`)
        .mockImplementation(() => mockPath)

      const errormessage = `this${mockPath}is\n${mockPath}a test: ${mockPath}`

      expect(cleanPaths(errormessage)).toBe(`this$PWDis\n$PWDa test: $PWD`)
      mockCwd.mockRestore()
    })
  })
})
