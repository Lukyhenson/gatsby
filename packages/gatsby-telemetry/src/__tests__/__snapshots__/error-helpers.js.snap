// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors Helpers sanitizeError Sanitizes current path from error stracktraces 1`] = `
"Object.it (src/__tests__/error-helpers.js:39:15)
Object.asyncJestTest (/Users/sidharthachatterjee/Code/gatsby/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:100:37)
resolve (/Users/sidharthachatterjee/Code/gatsby/node_modules/jest-jasmine2/build/queueRunner.js:41:12)
new Promise (<anonymous>)
mapper (/Users/sidharthachatterjee/Code/gatsby/node_modules/jest-jasmine2/build/queueRunner.js:26:19)
promise.then (/Users/sidharthachatterjee/Code/gatsby/node_modules/jest-jasmine2/build/queueRunner.js:71:41)
"
`;
